packages/upload-api/src/blob/add.js

```diff
@@ -0,0 +1,71 @@
+import * as Server from '@ucanto/server'
+import * as Blob from '@web3-storage/capabilities/blob'
+import * as API from '../types.js'
+
+/**
+ * @param {API.BlobServiceContext} context
+ * @returns {API.ServiceMethod<API.BlobAdd, API.BlobAddSuccess, API.BlobAddFailure>}
+ */
+export function blobAddProvider(context) {
+  return Server.provideAdvanced({
+    capability: Blob.add,
+    handler: async ({ capability, invocation }) => {
+      const { id, allocationStorage, maxUploadSize } = context
+      const { blob } = capability.nb
+      const space = /** @type {import('@ucanto/interface').DIDKey} */ (
+        Server.DID.parse(capability.with).did()
+      )
+
+      if (blob.size > maxUploadSize) {
+        return {
+          error: new Server.Failure(
+            `Maximum size exceeded: ${maxUploadSize}, split DAG into smaller shards.`
+          ),
+        }
+      }
+      
+      // Create effects for receipt
+      const [allocatefx, acceptfx] = await Promise.all([
+        Blob.allocate
+          .invoke({
+            issuer: id,
+            audience: id,
+            with: id.toDIDKey(),
+            nb: {
+              blob,
+              cause: invocation.link(),
+              space
+            },
+            expiration: Infinity,
+          })
+          .delegate(),
```

Comment:
Grabbing comment from @Gozala on https://github.com/web3-storage/w3up/pull/1340#discussion_r1536145607

I was thinking this one could be actual invocation that we run from here, I do not believe we need to schedule it for later. In my mind it would basically do what we have here


https://github.com/web3-storage/w3up/blob/9aba8a155cca663387952fc9b11bc3748de8f2fc/packages/upload-api/src/store/add.js#L13-L74

Than we would do what you do here. Followed by logic that checks if we already have blob, if so we could also issue receipt for accept. If we do not have it yet then do exactly what you do here.

Reply:
Grabbing comment from @Gozala on https://github.com/web3-storage/w3up/pull/1340#discussion_r1536150410

I'm not sure if we have a good way to check if we already have a receipt for this task or not, but assuming we can do that I imagine we'd have a big if clause around all these so if we do have a receipt we just check if blob was received and if so issue receipt for the accept if we do not have receipt yet it's a first run so we do what was described prior

Reply:
@Gozala looks like we are actually more closely aligned than I thought based on previous [code snipped shared in spec](https://github.com/web3-storage/specs/pull/115#discussion_r1535858479). This was as I understood an example on more futuristic Invocation spec, and not what we can do today or we would need to block on accept.

1. Starting by the first comment as adopting same strategy we had today in `store/add` handler with invocation right away instead of scheduling it for later, YES we can do as before, but I want to give more details on why I opted for this implementation for your understanding.
  1.1. In the future, I expect us to have extra step that would come in place during the scheduling phase. So, imagine moving from a world that handler has full knowledge right away on where to write (S3, R2 configured/hard coded) to a world where we have a protocol where `blob/*` service interacts with a protocol for discovery of write targets based on a set of heuristics in order to find out a presigned URL for the blob. I would expect this to happen async and in the service schedule, and therefore opting to go this way already.  @Gozala what do you think?
2. Regarding second comment and lookup for receipts, I still want to explore a bit more, but probably worth deciding after we align on first point here

Reply:
Me and @vasco-santos talked about this sync, but posting here for future reference.

Generally speaking all tasks should be async by nature, and only reason we had to resort to workarounds in filecoin pipeline because our client expected to get receipt as response. It is probably best if we try to keep working around client implementation detail and just make async receipts possible.

Apart from the client there is also question of the how do we proceed with a handlers on server. I think it would make sense to create generic invocation queue which can be basis for the scheduler. More concretely invocation CARs should be written in S3 (ideally in target space) and invocation CID put into invocation queue. Queue consumer can then

1. Check if there is already a receipt for this invocation, if so ignore and move on.
2. Otherwise fetch invocation, decode and dispatch to an appropriate handler
3. Write receipt into S3 (ideally in target space)

Client will only have to learn to poll receipts if one was not included in the response.

⚠️ It is important to properly handle concurrency however. Specifically queue consumer may check that receipt for the invocation does not exist and decide to run the task at the same time same task may be already running concurrently which would create a race. Which is to say I we need to write something before we run task to prevent this.

Reply:
> ⚠️ It is important to properly handle concurrency however. Specifically queue consumer may check that receipt for the invocation does not exist and decide to run the task at the same time same task may be already running concurrently which would create a race. Which is to say I we need to write something before we run task to prevent this.

The task result should be the same right, so it is not critical to check this but an optimization right?

Reply:
Note that @Gozala recommended to not do the scheduling and actually call `allocate` within `blob/add` for the time being. We can complicate things when we need a more async write target selection